src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonSerializerOptions.Caching.cs

```diff
@@ -157,39 +162,179 @@ internal void ClearCaches()
         /// </summary>
         internal sealed class CachingContext
         {
-            private readonly ConcurrentDictionary<Type, JsonTypeInfo?> _jsonTypeInfoCache = new();
+            private readonly ConcurrentDictionary<Type, CacheEntry> _cache = new();
 #if !NETCOREAPP
-            private readonly Func<Type, JsonTypeInfo?> _jsonTypeInfoFactory;
+            private readonly Func<Type, CacheEntry> _cacheEntryFactory;
 #endif
 
             public CachingContext(JsonSerializerOptions options, int hashCode)
             {
                 Options = options;
                 HashCode = hashCode;
 #if !NETCOREAPP
-                _jsonTypeInfoFactory = Options.GetTypeInfoNoCaching;
+                _cacheEntryFactory = type => CreateCacheEntry(type, this);
 #endif
             }
 
             public JsonSerializerOptions Options { get; }
             public int HashCode { get; }
             // Property only accessed by reflection in testing -- do not remove.
             // If changing please ensure that src/ILLink.Descriptors.LibraryBuild.xml is up-to-date.
-            public int Count => _jsonTypeInfoCache.Count;
+            public int Count => _cache.Count;
+
+            public JsonTypeInfo? GetOrAddTypeInfo(Type type, bool fallBackToNearestAncestorType = false)
+            {
+                CacheEntry entry = GetOrAddCacheEntry(type);
+                return fallBackToNearestAncestorType && !entry.HasResult
```

Comment:
> But I think it's a reasonable trade-off over the current behaviour of simply refusing to handle statically undeclared types.

I'm not convinced of that. Things that would have produced a diagnosable and actionable error are now going to just silently do the "wrong thing" in some cases.

Reply:
I'm mostly concerned about:

```csharp
using System.Text.Json;
using System.Text.Json.Serialization;

JsonSerializerOptions options = new()
{
    TypeInfoResolver = MyContext.Default,
};

Console.WriteLine(JsonSerializer.Serialize(new SomeData(), options));

[JsonSerializable(typeof(Person))]
partial class MyContext : JsonSerializerContext { }

class SomeData
{
    public object Data { get; set; } = new Customer() { Name = "John Doe", CompanyName = "Some Company" };
}

class Person
{
    public string Name { get; set; }
}

class Customer : Person
{
    public string CompanyName { get; set; }
}
```

Currently that throws

Reply:
> I'm not convinced of that. Things that would have produced a diagnosable and actionable error are now going to just silently do the "wrong thing" in some cases.

I would say that this fallback creates a loss of fidelity in all cases, not just a few cases. Even in the desirable scenario of, say, a private collection falling back to the contract for `IEnumerable<T>`, we are silently permitting serialization even though the type cannot be roundtripped to its original value (reflection permits that scenario). Still, I think this behaviour is desirable and removes a large hurdle in the adoption of SG, provided of course that we document it thoroughly (i.e. "In SG, serialization has contravariance semantics")

Even in the case of POCOs though, note that the "wrong thing" concerns serializing fewer properties than otherwise contained in the runtime type. STJ does this a in quite a few places already -- we're generally speaking more concerned about the opposite effect, i.e. unconstrained polymorphism serializing more properties than expected. 

I see two possible alternatives:

1. Ship as an opt-in flag in `JsonSerializerOptions` or
2. Restrict to private runtime types falling back to declared interface types, or something along these lines.

For Preview3, I would recommend shipping as-is to give a bit of soak time to the ASP.NET folks and get feedback for the feature in practice.

cc @eerhardt @davidfowl @halter73